File: app/src/main/java/com/example/HappensBefore2.java
package com.example;

public class HappensBefore2 {
  int x = 0;
  boolean done = false;

  public void thread1() {
    x = 1;
    done = true;
  }

  public void thread2() {
    while (!done) {
      //...
    }
    System.out.println(x);
  }
}

File: app/src/main/java/com/example/HappensBefore2.kt
package com.example

var xx = 0
var done = false

fun main() {

  val t1 = Thread {
    xx = 1
    done = true
  }

  val t2 = Thread {
    while (!done) {
      //...
    }
    println(xx)
  }

  t1.start()
  t2.start()
  t1.join()
  t2.join()
}